src/hotspot/share/services/nmtPreInit.hpp

```diff
@@ -143,22 +143,25 @@ class NMTPreInitAllocationTable {
   //  VMs with insanely long command lines maybe ~700-1000. Which gives us an expected
   //  load factor of ~.1. Hash collisions should be very rare.
   // ~8000 entries cost us ~64K for this table (64-bit), which is acceptable.
-  static const int table_size = 7919;
+  // We chose 8191, as this is a Mersenne prime (2^x - 1), which for a random
+  //  polynomial modulo p = (2^x - 1) is uniformily distributed in [p], so each
+  //  bit has the same distribution.
+  static const int table_size = 8191; // i.e. 8191==(2^13 - 1);
 
   NMTPreInitAllocation* _entries[table_size];
 
   typedef int index_t;
   const index_t invalid_index = -1;
 
   static unsigned calculate_hash(const void* p) {
-    uintptr_t tmp = p2i(p);
-    unsigned hash = (unsigned)tmp
-                     LP64_ONLY( ^ (unsigned)(tmp >> 32));
-    return hash;
+    // Keep hash function simple, the modulo
+    // operation in index function will do the "heavy lifting".
+    return (unsigned)p2i(p);
```

Comment:
How can the modulo do the "heavy lifting" if you shear off half of the value space? (well, about a quarter to a third, since most kernels have 48..52 bit addresses.

In fact, I would go the other direction and improve this hash like this:
return (tmp >> 3) ^ (tmp >> 35)

since the lower 3 bits are always 0 for malloc pointers.

Reply:
> How can the modulo do the "heavy lifting" if you shear off half of the value space? (well, about a quarter to a third, since most kernels have 48..52 bit addresses.
> 
> In fact, I would go the other direction and improve this hash like this: return (tmp >> 3) ^ (tmp >> 35)
> 
> since the lower 3 bits are always 0 for malloc pointers.

I ported the code wrong from my sandbox. The hash function should be dealing with 64 bits, not 32 bits:

```
  static unsigned calculate_hash(const void* p) {
    // Keep hash function simple, the modulo
    // operation in index function will do the "heavy lifting".
    return (unsigned)p2i(p);
  }

  static index_t index_for_key(const void* p) {
    const uint64_t hash = calculate_hash(p);
    // "table_size" is a Mersenne prime, so "modulo" is all we need here.
    return hash % table_size;
  }
```

The problem with your approach is that you are trying to extract "good" bits, by hand. I ran multiple experiments trying to find "uniform" bits by hand and it was hard. I never gotten a better hash compared to using full 64 bits, and using modulo operation (with Mersenne prime table size) to mix those 64 bits for us.

I will try your proposed hash function and report bak here.

Good catch about the 64 to 32 bit clipping - thank you.